smithy-model/src/main/java/software/amazon/smithy/model/validation/AbstractValidator.java

```diff
@@ -28,48 +28,106 @@ public String getName() {
+            FromSourceLocation location,
+            String message,
+            String... additionalEventIdParts
+    ) {
+        return createEvent(Severity.WARNING, shape, location, message, additionalEventIdParts);
     }
 
-    protected final ValidationEvent note(Shape shape, String message) {
-        return createEvent(Severity.NOTE, shape, shape.getSourceLocation(), message);
+    protected final ValidationEvent note(
+            Shape shape,
+            String message,
+            String... additionalEventIdParts
+    ) {
+        return createEvent(Severity.NOTE, shape, shape.getSourceLocation(), message, additionalEventIdParts);
     }
 
-    protected final ValidationEvent note(Shape shape, FromSourceLocation location, String message) {
-        return createEvent(Severity.NOTE, shape, location, message);
+    protected final ValidationEvent note(
+            Shape shape,
+            FromSourceLocation location,
+            String message,
+            String... additionalEventIdParts
+    ) {
+        return createEvent(Severity.NOTE, shape, location, message, additionalEventIdParts);
     }
 
-    protected final ValidationEvent createEvent(Severity severity, Shape shape, String message) {
-        return createEvent(severity, shape, shape.getSourceLocation(), message);
+    protected final ValidationEvent createEvent(
+            Severity severity,
+            Shape shape,
+            String message,
+            String... additionalEventIdParts
+    ) {
+        return createEvent(severity, shape, shape.getSourceLocation(), message, additionalEventIdParts);
     }
 
-    protected final ValidationEvent createEvent(Severity severity, Shape shape, FromSourceLocation loc, String msg) {
-        return createEvent(ValidationEvent.builder().severity(severity).message(msg)
-                                   .shapeId(shape.getId()).sourceLocation(loc.getSourceLocation()));
+    protected final ValidationEvent createEvent(
+            Severity severity,
+            Shape shape,
+            FromSourceLocation loc,
+            String msg,
+            String... additionalEventIdParts
+    ) {
+        return ValidationEvent.builder()
+                .severity(severity)
+                .message(msg)
+                .shapeId(shape.getId())
+                .sourceLocation(loc.getSourceLocation())
+                .id(assembleFullEventId(additionalEventIdParts))
+                .build();
     }
 
-    private ValidationEvent createEvent(ValidationEvent.Builder builder) {
-        return builder.id(getName()).build();
+    protected final String assembleFullEventId(String... additionalEventIdParts) {
+        String eventId = getName();
+        if (additionalEventIdParts.length == 1) {
+            eventId += "." + additionalEventIdParts[0];
+        } else if (additionalEventIdParts.length > 1) {
+            eventId += "." + String.join(".", additionalEventIdParts);
+        }
+        return eventId;
```

Comment:
Probably minor but, `String.join` does a lot more than what we need.

```suggestion
    if (additionalEventIdParts.length > 0) {
        StringBuilder buf = new StringBuilder(getName());
        for (String part : additionalEventIdParts) {
            buf.append('.').append(part);
        }
        return buf.toString();
    }
    return getName();
```

Reply:
Actually I can probably just use lines 126-130 in your suggested change in fact 👍